bundle/config/validate/folder_permissions.go

```diff
@@ -0,0 +1,135 @@
+package validate
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/databricks/cli/bundle"
+	"github.com/databricks/cli/bundle/libraries"
+	"github.com/databricks/cli/bundle/permissions"
+	"github.com/databricks/cli/libs/diag"
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/service/workspace"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/singleflight"
+)
+
+type folderPermissions struct {
+}
+
+// Apply implements bundle.ReadOnlyMutator.
+func (f *folderPermissions) Apply(ctx context.Context, b bundle.ReadOnlyBundle) diag.Diagnostics {
+	if len(b.Config().Permissions) == 0 {
+		return nil
+	}
+
+	rootPath := b.Config().Workspace.RootPath
+	paths := []string{}
+	if !libraries.IsVolumesPath(rootPath) {
+		paths = append(paths, rootPath)
+	}
+
+	if !strings.HasSuffix(rootPath, "/") {
+		rootPath += "/"
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.ArtifactPath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config().Workspace.ArtifactPath) {
+		paths = append(paths, b.Config().Workspace.ArtifactPath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.FilePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config().Workspace.FilePath) {
+		paths = append(paths, b.Config().Workspace.FilePath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.StatePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config().Workspace.StatePath) {
+		paths = append(paths, b.Config().Workspace.StatePath)
+	}
+
+	if !strings.HasPrefix(b.Config().Workspace.ResourcePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config().Workspace.ResourcePath) {
+		paths = append(paths, b.Config().Workspace.ResourcePath)
+	}
+
+	var diags diag.Diagnostics
+	g, ctx := errgroup.WithContext(ctx)
+	results := make([]diag.Diagnostics, len(paths))
+	syncGroup := new(singleflight.Group)
+	for i, p := range paths {
+		g.Go(func() error {
+			diags, err, _ := syncGroup.Do(p, func() (any, error) {
+				diags := checkFolderPermission(ctx, b, p)
```

Comment:
These paths are different by design so there won't be reuse at this level. I figured you could initialize the sync group here and pass it to `checkFolderPermissions`. Then inside of `getClosestExistingObject`, you can use `syncGroup.Do()` to make sure you're only doing a single call for any given path at any level.

Those calls will be done multiple times in the current approach.

Reply:
On second thought, this will apply only to cases where the different paths are not a child of the root. This should be infrequent enough that we can skip caching altogether.